server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+            origin_dest_prefix = self.get_config().get("remote_rel_prefix")
+        elif len(self.get_config().get("content_types")) >= 2:
+            origin_dest_prefix = os.path.join("content", "units")
+        else:
+            origin_dest_prefix = os.path.join("content", "units",
+                                              self.get_config().get("content_types")[0])
+
+        return origin_dest_prefix
+
+    def get_original_list(self):
+        """
+        Returns a list of files that need to be rsynced to the content directory
+
+        :return: List of files from /var/lib/pulp/content that need to be rsynced
+        :rtype: list of str
+        """
+        original_list = reduce(lambda x, y: x + y, self.original_lists.values())[:]
+        original_rel_prefix = misc.common_path(original_list)
+        for index, value in enumerate(original_list):
+            original_list[index] = os.path.relpath(value, original_rel_prefix)
+        return original_list
+
+    def get_original_prefix(self):
+        """
+        Returns the path to the directory in which Pulp stores the content units
+
+        :return: path to the content units
+        :rtype: str
+        """
+        original_list = reduce(lambda x, y: x + y, self.original_lists.values())
+        return misc.common_path(original_list)
+
+    def get_link_list(self, master, *args):
+        """
+        Returns list of symlinks that need to be coppied using rsync
+
+        :return: list of symlinks
+        :rtype: list of str
+        """
+        link_rel_prefix = misc.common_path(self.link_lists[master])
+        link_list = self.link_lists[master][:]
+        for p in range(len(link_list)):
+            link_list[p] = os.path.relpath(link_list[p], link_rel_prefix)
+        return link_list
+
+    def get_link_prefix(self, master_dir):
+        """
+        Returns relative path inside the working directory to where the symlinks are stored
+
+        :param master_dir: path to the master publish directory for the repo
+        :return: path to symlinks relative to the working directory
+        :rtype: str
+        """
+        return misc.common_path(self.link_lists[master_dir])
+
+    def _add_necesary_steps(self, date_filter=None, config=None):
+        """
+        This method needs to be implemented in each plugin. This method should include calls to
+        self.add_child() with particular rsync steps that are needed to perform the full publish
+        to the remote server.
+        """
+        raise NotImplementedError()
+
+    def _get_root_publish_dir(self):
+        """
+        Returns the publish directory path for the predistributor
+        """
+        raise NotImplementedError()
+
+    def _get_predistributor(self):
+        """
+        Returns distributor which repo has to be published in before
+        publish in rsync distributor, content generator.
+        """
+        raise NotImplementedError()
+
+
+class Lazy(object):
```

Comment:
Based on IRC discussion, we can remove this class and just organize the data differently that gets shared between steps.